lib/internal/modules/esm/loader.js

```diff
@@ -225,6 +231,39 @@ class ESMLoader {
     };
   }
 
+  /**
+   * Returns the url to use for the resolution of a given cache key url
+   * These are not guaranteed to be the same.
+   *
+   * In WHATWG HTTP spec for ESM the cache key is the non-I/O bound
+   * synchronous resolution using only string operations
+   *   ~= resolveImportMap(new URL(specifier, importerHREF))
+   *
+   * The url used for subsequent resolution is the response URL after
+   * all redirects have been resolved.
+   *
+   * https://example.com/foo redirecting to https://example.com/bar
+   * would have a cache key of https://example.com/foo and baseURL
+   * of https://example.com/bar
+   *
+   * MUST BE SYNCHRONOUS for import.meta initialization
+   * MUST BE CALLED AFTER receiving the url body due to I/O
+   * @param {string} url
+   * @returns {string}
+   */
+  baseURL(url) {
+    if (
+      StringPrototypeStartsWith(url, 'http:') ||
+      StringPrototypeStartsWith(url, 'https:')
+    ) {
+      // The request & response have already settled, so they are in
+      // fetchModule's cache, in which case, fetchModule returns
+      // immediately and synchronously
+      url = fetchModule(new URL(url), { parentURL: url }).resolvedHREF;
```

Comment:
Can we add an assertion here that URL is a string / URL and not undefined if a promise happens to be returned by `fetchModule`? Just feels like an important assertion to get into code and not just comments.

Reply:
@guybedford there is already a comment in fetch_module explaining that the caching works the way it does (promise -> cache actual value) specifically for this sort of use case but sure I agree.

Reply:
added